File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        
        maxVal = prices[-1]
        profit = 0
        for i in range(len(prices)-2,-1,-1):
            if prices[i] < maxVal:
                profit = max(profit,maxVal-prices[i])
            else:
                maxVal = prices[i]
        return profit
        

File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int curr_max = prices[n - 1];
        int ans = 0;
        for (int i = n-2; i>=0 ;i--){
            if (prices[i] < curr_max){
                ans = Math.max(ans,curr_max - prices[i]);
            }
            if (prices[i] > curr_max)
                curr_max = prices[i] ; 
        }
        return ans ; 
    }
}